packages/app/android/src/reactnative/java/io/invertase/firebase/common/ReactNativeFirebaseModule.java

```diff
@@ -74,19 +81,65 @@ public ReactContext getContext() {
   }
 
   public ExecutorService getExecutor() {
-    ExecutorService existingSingleThreadExecutor = executors.get(getName());
-    if (existingSingleThreadExecutor != null) return existingSingleThreadExecutor;
-    ExecutorService newSingleThreadExecutor = Executors.newSingleThreadExecutor();
-    executors.put(getName(), newSingleThreadExecutor);
-    return newSingleThreadExecutor;
+    return getExecutor(false, "");
   }
 
+  public ExecutorService getTransactionalExecutor() {
+    return getExecutor(true, "");
+  }
+
+  public ExecutorService getTransactionalExecutor(String identifier) {
+    return getExecutor(true, identifier);
+  }
+
+  public ExecutorService getExecutor(boolean isTransactional, String identifier) {
+    String executorName = getExecutorName(isTransactional, identifier);
+    ExecutorService existingExecutor = executors.get(executorName);
+    if (existingExecutor != null) return existingExecutor;
+    ExecutorService newExecutor = getNewExecutor(isTransactional);
+    executors.put(executorName, newExecutor);
+    return newExecutor;
+  }
+
+  private ExecutorService getNewExecutor(boolean isTransactional) {
+    if (isTransactional == true) {
+      return Executors.newSingleThreadExecutor();
+    } else {
+      ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(0, MAXIMUM_POOL_SIZE, KEEP_ALIVE_SECONDS, TimeUnit.SECONDS, new SynchronousQueue<Runnable>());
+      threadPoolExecutor.setRejectedExecutionHandler(executeInFallback);
+      return threadPoolExecutor;
+    }
+  }
+
+  private RejectedExecutionHandler executeInFallback = new RejectedExecutionHandler() {
+    public void rejectedExecution(Runnable r, ThreadPoolExecutor executor) {
+      ExecutorService fallbackExecutor = getTransactionalExecutor("");
+      fallbackExecutor.execute(r);
+    };
+  };
```

Comment:
The duplication between this and UniversalFirebaseModule needs a solution - having this vital logic duplicated makes no sense to me as I look at it. Is there any way to have the logic in Universal and just used by ReactNative ?

On a quick read of the code I don't see a way around it using inheritance or object access - they appear quite distinct :thinking: 

Maybe split the implementation itself out into a task  handling Util or something...